File: packages/cordova-consent/src/android/Consent.java
package cordova.plugin.consent;

import android.util.Log;
import android.util.SparseArray;

import com.google.android.ump.ConsentForm;
import com.google.android.ump.ConsentInformation;
import com.google.android.ump.ConsentRequestParameters;
import com.google.android.ump.UserMessagingPlatform;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;

import cordova.plugin.consent.Generated.Actions;
import cordova.plugin.consent.Generated.ConsentStatus;

public class Consent extends CordovaPlugin {
    private static final SparseArray<ConsentForm> forms = new SparseArray<ConsentForm>();
    private final ArrayList<PluginResult> eventQueue = new ArrayList<PluginResult>();
    private final String TAG = this.getClass().getSimpleName();
    private CallbackContext readyCallbackContext = null;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);

        ExecuteContext.plugin = this;
    }

    @Override
    public boolean execute(String actionKey, JSONArray args, final CallbackContext callbackContext) {
        ExecuteContext ctx = new ExecuteContext(actionKey, args, callbackContext);
        Log.d(TAG, actionKey);

        switch (actionKey) {
            case Actions.READY:
                return executeReady(callbackContext);
            case Actions.GET_CONSENT_STATUS:
                callbackContext.success(getConsentStatus());
                break;
            case Actions.GET_FORM_STATUS:
                callbackContext.success(getConsentInformation().isConsentFormAvailable() ? 1 : 2);
                break;
            case Actions.REQUEST_INFO_UPDATE:
                return executeRequestInfoUpdate(ctx);
            case Actions.LOAD_FORM:
                return executeLoadForm(ctx);
            case Actions.SHOW_FORM:
                return executeShowForm(ctx);
            case Actions.RESET:
                getConsentInformation().reset();
                callbackContext.success();
                break;
            default:
                return false;
        }

        return true;
    }

    private int getConsentStatus() {
        int status = getConsentInformation().getConsentStatus();
        switch (status) {
            case ConsentInformation.ConsentStatus.NOT_REQUIRED:
                return ConsentStatus.NOT_REQUIRED;
            case ConsentInformation.ConsentStatus.REQUIRED:
                return ConsentStatus.REQUIRED;
            default:
                return status;
        }
    }

    private boolean executeReady(CallbackContext callbackContext) {
        if (readyCallbackContext == null) {
            for (PluginResult result : eventQueue) {
                callbackContext.sendPluginResult(result);
            }
            eventQueue.clear();
        } else {
            Log.e(TAG, "Ready action should only be called once.");
        }
        readyCallbackContext = callbackContext;
        emit(Generated.Events.READY);
        return true;
    }

    private boolean executeRequestInfoUpdate(ExecuteContext ctx) {
        ConsentRequestParameters params = ctx.optConsentRequestParameters();
        ConsentInformation consentInformation = getConsentInformation();
        consentInformation.requestConsentInfoUpdate(
                cordova.getActivity(),
                params,
                ctx.callbackContext::success,
                formError -> ctx.callbackContext.error(formError.getMessage()));

        return true;
    }

    private boolean executeLoadForm(ExecuteContext ctx) {
        cordova.getActivity().runOnUiThread(() -> {
            UserMessagingPlatform.loadConsentForm(
                    cordova.getActivity(),
                    consentForm -> {
                        forms.put(consentForm.hashCode(), consentForm);
                        ctx.callbackContext.success(consentForm.hashCode());
                    },
                    formError -> ctx.callbackContext.error(formError.getMessage())
            );
        });
        return true;
    }

    private boolean executeShowForm(ExecuteContext ctx) {
        final ConsentForm consentForm = forms.get(ctx.optId());
        cordova.getActivity().runOnUiThread(() -> {
            consentForm.show(
                    cordova.getActivity(),
                    formError -> {
                        if (formError == null) {
                            ctx.callbackContext.success();
                        } else {
                            ctx.callbackContext.error(formError.getMessage());
                        }
                    });
        });
        return true;
    }

    private ConsentInformation getConsentInformation() {
        return UserMessagingPlatform.getConsentInformation(cordova.getActivity());
    }

    @Override
    public void onDestroy() {
        readyCallbackContext = null;

        super.onDestroy();
    }

    public void emit(String eventType) {
        emit(eventType, null);
    }

    public void emit(String eventType, Object data) {
        JSONObject event = new JSONObject();
        try {
            event.put("type", eventType);
            event.put("data", data);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        PluginResult result = new PluginResult(PluginResult.Status.OK, event);
        result.setKeepCallback(true);
        if (readyCallbackContext == null) {
            eventQueue.add(result);
        } else {
            readyCallbackContext.sendPluginResult(result);
        }
    }
}

File: packages/cordova-consent/src/android/Consent.kt
package cordova.plugin.consent

import android.util.Log
import android.util.SparseArray
import com.google.android.ump.ConsentForm
import com.google.android.ump.ConsentInformation
import com.google.android.ump.FormError
import com.google.android.ump.UserMessagingPlatform
import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject


class Consent : CordovaPlugin() {
    private val eventQueue = ArrayList<PluginResult>()
    private val TAG = this.javaClass.simpleName
    private var readyCallbackContext: CallbackContext? = null

    private val actions = mapOf(
        Actions.READY to ::executeReady,
        Actions.GET_CONSENT_STATUS to ::executeGetConsentStatus,
        Actions.GET_FORM_STATUS to ::executeGetFormStatus,
        Actions.REQUEST_INFO_UPDATE to ::executeRequestInfoUpdate,
        Actions.LOAD_FORM to ::executeLoadForm,
        Actions.SHOW_FORM to ::executeShowForm,
        Actions.RESET to ::executeReset,
        Actions.CAN_REQUEST_ADS to ::executeCanRequestAds,
        Actions.LOAD_AND_SHOW_IF_REQUIRED to ::executeLoadAndShowIfRequired,
        Actions.SHOW_PRIVACY_OPTIONS_FORM to ::executeShowPrivacyOptionsForm,
        Actions.PRIVACY_OPTIONS_REQUIREMENT_STATUS to ::executePrivacyOptionsRequirementStatus,
    )

    override fun execute(
        action: String,
        args: JSONArray,
        callbackContext: CallbackContext
    ): Boolean {
        val ctx =
            ExecuteContext(action, args, callbackContext, this)
        Log.d(TAG, action)
        return actions[action]?.invoke(ctx) != null
    }

    private val consentStatus: Int
        get() {
            return when (val status = consentInformation.consentStatus) {
                ConsentInformation.ConsentStatus.NOT_REQUIRED -> ConsentStatus.NOT_REQUIRED
                ConsentInformation.ConsentStatus.REQUIRED -> ConsentStatus.REQUIRED
                else -> status
            }
        }

    private fun executeReady(ctx: ExecuteContext) {
        if (readyCallbackContext == null) {
            for (result in eventQueue) {
                ctx.callbackContext.sendPluginResult(result)
            }
            eventQueue.clear()
        } else {
            Log.e(TAG, "Ready action should only be called once.")
        }
        readyCallbackContext = ctx.callbackContext
        emit(Events.READY)
    }

    private fun executeCanRequestAds(ctx: ExecuteContext) {
        ctx.resolve(consentInformation.canRequestAds())
    }

    private fun executeLoadAndShowIfRequired(ctx: ExecuteContext) {
        cordova.activity.runOnUiThread {
            UserMessagingPlatform.loadAndShowConsentFormIfRequired(
                cordova.activity
            ) { loadAndShowError ->
                loadAndShowError?.let {
                    ctx.callbackContext.error(it.message)
                } ?: run {
                    ctx.resolve()
                }
            }
        }
    }

    private fun executeShowPrivacyOptionsForm(ctx: ExecuteContext) {
        cordova.activity.runOnUiThread {
            UserMessagingPlatform.showPrivacyOptionsForm(cordova.activity) { formError ->
                formError?.let {
                    ctx.callbackContext.error(it.message)
                } ?: run {
                    ctx.resolve()
                }
            }
        }
    }

    private fun executePrivacyOptionsRequirementStatus(ctx: ExecuteContext) {
        ctx.resolve(consentInformation.privacyOptionsRequirementStatus)
    }

    private fun executeGetConsentStatus(ctx: ExecuteContext) {
        ctx.callbackContext.success(consentStatus)
    }

    private fun executeGetFormStatus(ctx: ExecuteContext) {
        ctx.callbackContext.success(if (consentInformation.isConsentFormAvailable) 1 else 2)
    }

    private fun executeRequestInfoUpdate(ctx: ExecuteContext) {
        val params = ctx.optConsentRequestParameters()
        val consentInformation = consentInformation
        cordova.activity.runOnUiThread {
            consentInformation.requestConsentInfoUpdate(
                cordova.activity,
                params, { ctx.callbackContext.success() }
            ) { formError: FormError -> ctx.callbackContext.error(formError.message) }
        }
    }

    private fun executeLoadForm(ctx: ExecuteContext) {
        cordova.activity.runOnUiThread {
            UserMessagingPlatform.loadConsentForm(
                cordova.activity,
                { consentForm: ConsentForm ->
                    forms.put(consentForm.hashCode(), consentForm)
                    ctx.callbackContext.success(consentForm.hashCode())
                }
            ) { formError: FormError -> ctx.callbackContext.error(formError.message) }
        }
    }

    private fun executeShowForm(ctx: ExecuteContext) {
        val consentForm = forms[ctx.optId()]
        cordova.activity.runOnUiThread {
            consentForm.show(
                cordova.activity
            ) { formError: FormError? ->
                if (formError == null) {
                    ctx.callbackContext.success()
                } else {
                    ctx.callbackContext.error(formError.message)
                }
            }
        }
    }

    private fun executeReset(ctx: ExecuteContext) {
        consentInformation.reset()
        ctx.callbackContext.success()
    }

    private val consentInformation: ConsentInformation
        get() = UserMessagingPlatform.getConsentInformation(cordova.activity)

    override fun onDestroy() {
        readyCallbackContext = null
        super.onDestroy()
    }

    @JvmOverloads
    fun emit(eventType: String?, data: Any? = null) {
        val event = JSONObject()
        try {
            event.put("type", eventType)
            event.put("data", data)
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        val result = PluginResult(PluginResult.Status.OK, event)
        result.keepCallback = true
        if (readyCallbackContext == null) {
            eventQueue.add(result)
        } else {
            readyCallbackContext!!.sendPluginResult(result)
        }
    }

    companion object {
        private val forms = SparseArray<ConsentForm>()
    }
}
